nomad/csi_endpoint.go

```diff
@@ -410,13 +418,141 @@ func (v *CSIPlugin) Get(args *structs.CSIPluginGetRequest, reply *structs.CSIPlu
 func (srv *Server) controllerUnpublishVolume(req *structs.CSIVolumeClaimRequest, nodeID string) error {
-	// TODO(tgross): implement me!
+	plug, vol, err := srv.volAndPluginLookup(req.VolumeID)
+	if plug == nil || vol == nil || err != nil {
+		return err // possibly nil if no controller required
+	}
+
+	method := "ClientCSI.DetachVolume"
+	cReq := &cstructs.ClientCSIControllerDetachVolumeRequest{
+		PluginName: plug.ID,
+		VolumeID:   req.VolumeID,
+		NodeID:     nodeID,
+	}
+	err = srv.csiControllerRPC(plug, method, cReq,
+		&cstructs.ClientCSIControllerDetachVolumeResponse{})
+	if err != nil {
+		return err
+	}
 	return nil
 }
+
+func (srv *Server) volAndPluginLookup(volID string) (*structs.CSIPlugin, *structs.CSIVolume, error) {
+	state := srv.fsm.State()
+	ws := memdb.NewWatchSet()
+
+	vol, err := state.CSIVolumeByID(ws, volID)
+	if err != nil {
+		return nil, nil, err
+	}
+	if vol == nil {
+		return nil, nil, fmt.Errorf("volume not found: %s", volID)
+	}
+	if !vol.ControllerRequired {
+		return nil, nil, nil
+	}
+
+	// note: we do this same lookup in CSIVolumeByID but then throw
+	// away the pointer to the plugin rather than attaching it to
+	// the volume so we have to do it again here.
+	plug, err := state.CSIPluginByID(ws, vol.PluginID)
+	if err != nil {
+		return nil, nil, err
+	}
+	if plug == nil {
+		return nil, nil, fmt.Errorf("plugin not found: %s", vol.PluginID)
+	}
+	return plug, vol, nil
+}
+
+func (srv *Server) csiControllerRPC(plugin *structs.CSIPlugin, method string, args, reply interface{}) error {
+	// plugin IDs are not scoped to region/DC but volumes are.
+	// so any node we get for a controller is already in the same region/DC
+	// for the volume.
+	nodeID, err := srv.nodeForControllerPlugin(plugin)
+	if err != nil || nodeID == "" {
+		return err
+	}
+	err = findNodeConnAndForward(srv, nodeID, method, args, reply)
+	if err != nil {
+		return err
+	}
+	if replyErr, ok := reply.(error); ok {
+		return replyErr
+	}
+	return nil
+}
+
+// nodeForControllerPlugin returns the node ID for a random controller
+// to load-balance long-blocking RPCs across client nodes.
+func (srv *Server) nodeForControllerPlugin(plugin *structs.CSIPlugin) (string, error) {
+	count := len(plugin.Controllers)
+	if count == 0 {
+		return "", fmt.Errorf("no controllers available for plugin %q", plugin.ID)
+	}
+	snap, err := srv.fsm.State().Snapshot()
+	if err != nil {
+		return "", err
+	}
+
+	// iterating maps is "random" but unspecified and isn't particularly
+	// random with small maps, so not well-suited for load balancing.
+	// so we shuffle the keys and iterate over them.
+	clientIDs := []string{}
+	for clientID := range plugin.Controllers {
+		clientIDs = append(clientIDs, clientID)
+	}
+	rand.Shuffle(count, func(i, j int) {
+		clientIDs[i], clientIDs[j] = clientIDs[j], clientIDs[i]
+	})
+
+	for _, clientID := range clientIDs {
+		controller := plugin.Controllers[clientID]
+		ci := controller.ControllerInfo
```

Comment:
We might wanna define a helper on CSIInfo for `IsController() bool` and `IsNode() bool` because right now this would be confusing to future readers probably.

Reply:
Good call. Done and added some commentary to the use of the check here as well.

Reply:
Whoops and found a lurking null-pointer bug in the process! Fixed that too!